WalletWasabi/Helpers/PowerSaving/BaseInhibitorTask.cs

```diff
@@ -0,0 +1,173 @@
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+using WalletWasabi.Microservices;
+
+namespace WalletWasabi.Helpers.PowerSaving;
+
+public class BaseInhibitorTask : IPowerSavingInhibitorTask
+{
+	/// <remarks>Guarded by <see cref="StateLock"/>.</remarks>
+	private bool _isDone;
+
+	protected BaseInhibitorTask(TimeSpan period, string reason, ProcessAsync process)
+	{
+		BasePeriod = period;
+		Reason = reason;
+		Process = process;
+		Cts = new CancellationTokenSource(period);
+
+		_ = WaitAsync();
+	}
+
+	/// <remarks>Guards <see cref="_isDone"/>.</remarks>
+	protected object StateLock { get; } = new();
+
+	/// <inheritdoc/>
+	public bool IsDone
+	{
+		get
+		{
+			lock (StateLock)
+			{
+				return _isDone;
+			}
+		}
+	}
+
+	/// <remarks>It holds: inhibitorEndTime = now + BasePeriod + ProlongInterval.</remarks>
+	public TimeSpan BasePeriod { get; }
+
+	/// <summary>Reason why the power saving is inhibited.</summary>
+	public string Reason { get; }
+	private ProcessAsync Process { get; }
+	private CancellationTokenSource Cts { get; }
+	private TaskCompletionSource StoppedTcs { get; } = new();
+
+	private async Task WaitAsync()
+	{
+		try
+		{
+			await Process.WaitForExitAsync(Cts.Token).ConfigureAwait(false);
+
+			// This should be hit only when somebody externally kills the inhibiting process.
+			Logger.LogError("Inhibit process ended prematurely.");
+		}
+		catch (OperationCanceledException)
+		{
+			Logger.LogTrace("Elapsed time limit for the inhibitor task to live.");
+		}
```

Comment:
If there is another type of exception it can end up as unobserved. Consider adding general exception handler.